File: backend/app/services/calculator.py
import logging
from bisect import bisect_left
from decimal import Decimal
from app.services.constants import (
    BREAKPOINT_NOT_FOUND_ERROR,
    HIGH_INDEX,
    LOW_INDEX,
)
from app.core.exceptions import BreakpointNotFoundError
from app.core.telemetry import emit_event
from app.domain.entities import LoanApplication
from app.domain.interfaces import IFeeRepository, IFeeInterpolationStrategy
from app.domain.logic import round_to_multiple_of_five

class FeeCalculatorService:
    def __init__(
        self,
        repository: IFeeRepository,
        strategy: IFeeInterpolationStrategy,
    ):
        self.repository = repository
        self.strategy = strategy

    def calculate_fee(self, application: LoanApplication) -> Decimal:
        breakpoints = self.repository.get_breakpoints(application.term)
        is_sorted = all(
            breakpoints[index][LOW_INDEX] <= breakpoints[index + 1][LOW_INDEX]
            for index in range(len(breakpoints) - 1)
        )
        if not is_sorted:
            breakpoints = sorted(breakpoints, key=lambda point: point[LOW_INDEX])
        amounts = [point[LOW_INDEX] for point in breakpoints]
        position = bisect_left(amounts, application.amount)

        if position >= len(breakpoints) or (position == 0 and amounts[position] != application.amount):
            error = BreakpointNotFoundError(
                BREAKPOINT_NOT_FOUND_ERROR.format(term=application.term, amount=application.amount),
                details={
                    "term": application.term,
                    "amount": str(application.amount),
                },
            )
            emit_event(
                logging.ERROR,
                "service_calculate_fee_breakpoint_not_found",
                {"amount": str(application.amount), "term": application.term},
                error,
            )
            raise error

        if amounts[position] == application.amount:
            rounded_fee = round_to_multiple_of_five(application.amount, breakpoints[position][HIGH_INDEX])
            return rounded_fee

        lower = breakpoints[position - 1]
        upper = breakpoints[position]
        base_fee = self.strategy.calculate(application.amount, lower, upper)
        rounded_fee = round_to_multiple_of_five(application.amount, base_fee)
        return rounded_fee

File: backend-go/internal/services/calculator.go
package services

import (
	"fmt"
	"sort"

	"github.com/johnsonafuye/fee-calculator-backend-go/internal/domain"
	"github.com/shopspring/decimal"
)

type FeeCalculatorService struct {
	repository domain.FeeRepository
	strategy   domain.FeeInterpolationStrategy
}

func NewFeeCalculatorService(repo domain.FeeRepository, strategy domain.FeeInterpolationStrategy) *FeeCalculatorService {
	return &FeeCalculatorService{
		repository: repo,
		strategy:   strategy,
	}
}

func (s *FeeCalculatorService) CalculateFee(application *domain.LoanApplication) (decimal.Decimal, error) {
	breakpoints := s.repository.GetBreakpoints(application.Term)

	// In Go, since we use static slices, they are already sorted, but we reproduce the same robust check for safety.
	isSorted := true
	for i := 0; i < len(breakpoints)-1; i++ {
		if breakpoints[i].LowerBound.GreaterThan(breakpoints[i+1].LowerBound) {
			isSorted = false
			break
		}
	}
	if !isSorted {
		sort.Slice(breakpoints, func(i, j int) bool {
			return breakpoints[i].LowerBound.LessThan(breakpoints[j].LowerBound)
		})
	}

	amounts := make([]decimal.Decimal, len(breakpoints))
	for i, bp := range breakpoints {
		amounts[i] = bp.LowerBound
	}

	// bisect_left equivalent
	position := sort.Search(len(amounts), func(i int) bool {
		return amounts[i].GreaterThanOrEqual(application.Amount)
	})

	if position >= len(breakpoints) || (position == 0 && !amounts[position].Equal(application.Amount)) {
		return decimal.Zero, &domain.BreakpointNotFoundError{
			Message: fmt.Sprintf("Unable to find fee breakpoints for term=%d amount=%v", application.Term, application.Amount),
			Term:    application.Term,
			Amount:  application.Amount.String(),
		}
	}

	if amounts[position].Equal(application.Amount) {
		roundedFee := domain.RoundToMultipleOfFive(application.Amount, breakpoints[position].Fee)
		return roundedFee, nil
	}

	lower := breakpoints[position-1]
	upper := breakpoints[position]

	baseFee, err := s.strategy.Calculate(application.Amount, lower, upper)
	if err != nil {
		return decimal.Zero, err
	}

	roundedFee := domain.RoundToMultipleOfFive(application.Amount, baseFee)
	return roundedFee, nil
}
